pkg/config/namewatcher.go

```diff
@@ -0,0 +1,83 @@
+package config
+
+import (
+	"os"
+	"strings"
+
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/klog/v2"
+)
+
+const (
+	sharedSecretReservedNamesEnvVarName    = "RESERVED_SHARED_SECRET_NAMES"
+	sharedConfigMapReservedNamesEnvVarName = "RESERVED_SHARED_CONFIGMAP_NAMES"
+)
+
+type ReservedNames struct {
+	currentSecretNames    map[string]types.NamespacedName
+	currentConfigMapNames map[string]types.NamespacedName
+}
+
+func SetupNameReservation() *ReservedNames {
+	rn := &ReservedNames{}
+	sharedSecretReservations := os.Getenv(sharedSecretReservedNamesEnvVarName)
+	rn.currentSecretNames = parseEnvVar(sharedSecretReservations, sharedSecretReservedNamesEnvVarName)
+	klog.Infof("reserved shared secret env %s and map %#v", sharedSecretReservations, rn.currentSecretNames)
+	sharedConfigMapReservations := os.Getenv(sharedConfigMapReservedNamesEnvVarName)
+	rn.currentConfigMapNames = parseEnvVar(sharedConfigMapReservations, sharedConfigMapReservedNamesEnvVarName)
+	klog.Infof("reserved shared configmap env %s and map %#v", sharedConfigMapReservations, rn.currentConfigMapNames)
+	return rn
+}
+
+func parseEnvVar(val, envvarName string) map[string]types.NamespacedName {
+	ret := map[string]types.NamespacedName{}
+	entries := strings.Split(val, ";")
+	for _, entry := range entries {
+		a := strings.Split(entry, ":")
+		if len(a) < 3 {
+			if len(entry) > 0 {
+				klog.Warningf("env var %s has bad entry %s", envvarName, entry)
+			}
+			continue
+		}
+		nsName := types.NamespacedName{
+			Namespace: strings.TrimSpace(a[1]),
+			Name:      strings.TrimSpace(a[2]),
+		}
+		ret[strings.TrimSpace(a[0])] = nsName
+	}
+	return ret
+}
+
+func (rn *ReservedNames) ValidateSharedSecretOpenShiftName(shareName, refNamespace, refName string) bool {
+	v, ok := rn.currentSecretNames[shareName]
+	return innerValidate(shareName, refNamespace, refName, ok, v)
+}
+
+func (rn *ReservedNames) ValidateSharedConfigMapOpenShiftName(shareName, refNamespace, refName string) bool {
+	v, ok := rn.currentConfigMapNames[shareName]
+	return innerValidate(shareName, refNamespace, refName, ok, v)
+}
+
+func startsWithOpenShift(shareName string) bool {
+	if strings.HasPrefix(shareName, "openshift-") ||
+		strings.HasPrefix(shareName, "openshift_") {
+		return true
+	}
```

Comment:
I condensed this to just one `HasPrefix` call on `openshift` as part of the new unit tests

That said, the amount of shares being created is going to be small on a given cluster (single digits).  With that, I'm inclined to keep it simple (i.e. no regexp), and would rather keep things all in namewatcher and the new method, in case we decide to modify the reserve criteria in the future.

Reply:
I will defer to your decision based on the intended amount of usage.